frame/benchmarking/README.md

```diff
@@ -1,3 +1,182 @@
+of the function, for example triggering more iterations in a `for` loop, to get an accurate result,
+you must set up your benchmark to trigger this.
+
+It may be that there are multiple paths your function can go down, and it is not clear which one is
+most heavy. In this case, you should just create a benchmark for each scenario! You may find that
+there are paths in your code where complexity may become unbounded depending on user input. This may
+hint to you that you should implement sane boundaries for how a user can use your pallet. For
+example: limiting the number of elements in a vector, limiting the number of iterations in a `for`
+loop, etc...
+
+Examples of end to end benchmarks can be found in the [pallets provided by Substrate](../), and the
+specific details on how to use the `benchmarks!` macro can be found in [its
+documentation](./src/lib.rs).
+
+## Testing Benchmarks
+
+You can test your benchmarks using the same test runtime that you created for your pallet's unit
+tests. By creating your benchmarks in the `benchmarks!` macro, it automatically generates test
+functions for you:
+
+```rust
+fn test_benchmark_[benchmark_name]<T>::() -> Result<(), &'static str>
+```
+
+Simply add these functions to a unit test and ensure that the result of the function is `Ok(())`.
+
+> **Note:** If your test runtime and production runtime have different configurations, you may get
+different results when testing your benchmark and actually running it.
+
+In general, benchmarks returning `Ok(())` is all you need to check for since it signals the executed
+extrinsic has completed successfully. However, you can optionally include a `verify` block with your
+benchmark which can additionally verify any final conditions, such as the final state of your
+runtime.
+
+These additional `verify` blocks will not effect the results of your final benchmarking process.
+
+To run the tests, you need to be sure to enable the `runtime-benchmarks` feature flag. This may also
+mean you need to move into your node's binary folder. For example, with the Substrate repository,
+this is how you would test the Balances Pallet's benchmarks:
+
+```bash
+cd bin/node/cli
+cargo test -p pallet-balances --features runtime-benchmarks
+```
+
+## Adding Benchmarks
+
+The benchmarks included with each pallet are not automatically added to your node. To actually
+execute these benchmarks, you need to implement the `frame_benchmarking::Benchmark` trait. You can
+see an example of how to do this in the [included Substrate
+node](../../bin/node/runtime/src/lib.rs).
+
+Assuming there are already some benchmarks set up on your node, you just need to add another
+instance of the `add_benchmarks!` macro:
+
+```rust
+///  configuration for running benchmarks
+///               |    name of your pallet's crate (as imported)
+///               v                   v
+add_benchmark!(params, batches, pallet_balances, Balances);
```

Comment:
Should there be a `#[cfg(feature = "runtime_benchmarks")]` before this?

Reply:
ehh, its actually already wrapped in a runtime api which has this feature flag.

I have omitted a lot of code here for simplicity, but I think if you `find` this text in substrate, it would be clear what to do.

Reply:
you can let me know otherwise